File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        # Step 1: Split the string into words using split() which handles multiple spaces
        words = s.split()
        
        # Step 2: Reverse the list of words
        reversed_words = words[::-1]
        
        # Step 3: Join the reversed list of words with a single space
        return ' '.join(reversed_words)

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
#include <string>
#include <sstream>
#include <vector>
#include <algorithm>

class Solution {
public:
    std::string reverseWords(std::string s) {
        // Step 1: Trim leading and trailing spaces
        int left = 0, right = s.size() - 1;
        while (left <= right && s[left] == ' ') left++;
        while (right >= left && s[right] == ' ') right--;
        
        // Step 2: Split words and ignore multiple spaces
        std::vector<std::string> words;
        std::string word;
        while (left <= right) {
            char c = s[left];
            if (c != ' ') {
                word += c;
            } else if (!word.empty()) {
                words.push_back(word);
                word.clear();
            }
            left++;
        }
        if (!word.empty()) {
            words.push_back(word);
        }

        // Step 3: Reverse the order of words
        std::reverse(words.begin(), words.end());
        
        // Step 4: Join the words with a single space
        std::string result;
        for (int i = 0; i < words.size(); ++i) {
            result += words[i];
            if (i < words.size() - 1) {
                result += " ";
            }
        }
        
        return result;
    }
};
